### ISSUE TITLE
sodium_crypto_secretbox() expects parameter 1 to be string, array given

### ISSUE THREAD

**Issue by tjaydk** [pusher/pusher-http-php#290]
Hi.

Im currently working on implementing a direct chat and therefore i want messages to be encrypted. Everything has worked with private channels up until trying to use the PrivateEncrypted channel. Im using Laravel 8+ and PHP 7.4.

**Broadcasting.php**
```
'pusher' => [
            'driver' => 'pusher',
            'key' => env('PUSHER_APP_KEY'),
            'secret' => env('PUSHER_APP_SECRET'),
            'app_id' => env('PUSHER_APP_ID'),
            'options' => [
                'cluster' => env('PUSHER_APP_CLUSTER'),
                'useTLS' => true,
                'encryption_master_key_base64' => env('PUSHER_APP_ENC_MASTER_KEY')
            ],
        ],
```
**ChatChannel.php**
```
public function join(JwtUserDTO $jwt, int $chatId)
    {
        $chat = $this->chatService->getChatByChatId($chatId);
        if ($jwt->getUserId() !== $chat->getUserIdFrom() && $jwt->getUserId() !== $chat->getUserIdTo()) {
            return false;
        }
        return true;
    }
```
**ChatMessageCreateEvent.php**
```
public function broadcastOn()
    {
        return new EncryptedPrivateChannel('chat.' . $this->chatMessage->getChatId());
    }
```

When posting a new chat message i get the following exception
```
TypeError
sodium_crypto_secretbox() expects parameter 1 to be string, array given
```
Triggered from:
**PusherCrypto.php - Line 119 ("pusher/pusher-php-server": "^5.0",)**
```
public function encrypt_payload($channel, $plaintext)
    {
        if (!self::is_encrypted_channel($channel)) {
            throw new PusherException('Cannot encrypt plaintext for a channel that is not of the form private-encrypted-*. Got '.$channel);
        }
        $nonce = $this->generate_nonce();
        $shared_secret = $this->generate_shared_secret($channel);
        $cipher_text = sodium_crypto_secretbox($plaintext, $nonce, $shared_secret);

        return $this->format_encrypted_message($nonce, $cipher_text);
    }
```

I am unsure why the plaintext payload is an array - and how i can change it to ensure that it is only a string?

**Comment by elverkilde** [comment]
Are you able to share the code you are using to publish? I assume you are using `pusher->trigger($channel, .., ..)` somewhere?

**Comment by tjaydk** [comment]
Hi @elverkilde 

Im using the BroadcastServiceProvider and then broadcast my event

**ChatService.php**
```
public function createMessage(ChatMessageDTO $chatMessage, int $chatId): ChatMessageDTO
    {
        $chatMessage = $this->chatAggregator->createMessage($chatMessage, $chatId);
        broadcast(new ChatMessageCreateEvent($chatMessage));
        return $chatMessage;
    }
```
**ChatMessageCreateEvent.php**
```
<?php

namespace App\Events\ChatEvents;

use App\DTOs\ChatDTOs\ChatMessageDTO;
use Illuminate\Broadcasting\EncryptedPrivateChannel;
use Illuminate\Broadcasting\InteractsWithSockets;
use Illuminate\Contracts\Broadcasting\ShouldBroadcast;
use Illuminate\Foundation\Events\Dispatchable;
use Illuminate\Queue\SerializesModels;

class ChatMessageCreateEvent implements ShouldBroadcast
{
    use Dispatchable, InteractsWithSockets, SerializesModels;

    public ChatMessageDTO $chatMessage;

    /**
     * Create a new event instance.
     *
     * @param ChatMessageDTO $chatMessage
     */
    public function __construct(ChatMessageDTO $chatMessage)
    {
        $this->chatMessage = $chatMessage;
    }

    /**
     * @inheritDoc
     */
    public function broadcastOn()
    {
        return new EncryptedPrivateChannel('chat.' . $this->chatMessage->getChatId());
    }

    /**
     * @return string
     */
    public function broadcastAs()
    {
        return 'chat.message.new';
    }
    
//    Able to manipulate the data - but it is still passed as an array
//    public function broadcastWith()
//    {
//        return ['$data' => 'hello'];
//    }
}

```
This uses the PusherBroadcaster which indeed calls the `pusher->trigger(...)`
```
public function broadcast(array $channels, $event, array $payload = [])
    {
        $socket = Arr::pull($payload, 'socket');

        $response = $this->pusher->trigger(
            $this->formatChannels($channels), $event, $payload, $socket, true
        );

        if ((is_array($response) && $response['status'] >= 200 && $response['status'] <= 299)
            || $response === true) {
            return;
        }

        throw new BroadcastException(
            ! empty($response['body'])
                ? sprintf('Pusher error: %s.', $response['status'])
                : 'Failed to connect to Pusher.'
        );
    }
```
Unfortunately this has an 'already_encoded' hardcoded value of true - which means that the json_encode is not triggered and the whole array is passed to the `encrypt_payload($channel, $plaintext)`

By manipulating the `broadcastWith` function in the event im able to customize the payload, but it is always passed to the function as an array. And I can't find where to change the hardcoded 'already_encoded' value.

**Comment by elverkilde** [comment]
Huh. That definitely seems to be a bug, I don't see how that would ever work (feeding an array into a function assuming a JSON string) 🤔 

Forgive my ignorance, where is this broadcast function coming from -- Laravel?

**Comment by tjaydk** [comment]
Yes, it is from Laravel.

**Comment by elverkilde** [comment]
This seems relevant: https://github.com/illuminate/broadcasting/blob/master/Broadcasters/PusherBroadcaster.php#L114-L120

Is this the version of the broadcaster you are running? Is your version of Pusher greater than 5?

**Comment by tjaydk** [comment]
Im running `pusher-php-server: 5.0.1`
```
"name": "pusher/pusher-php-server",
            "version": "v5.0.1",
            "source": {
                "type": "git",
                "url": "https://github.com/pusher/pusher-http-php.git",
                "reference": "6cad1269f1f5261969848b72829f4f30736f3d90"
            },
```
And I have just tried to update to `laravel/framework: 8.10.0`
```
"name": "laravel/framework",
            "version": "v8.10.0",
            "source": {
                "type": "git",
                "url": "https://github.com/laravel/framework.git",
                "reference": "0c80950806cd1bc6d9a7068585a12c2bfa23bdf3"
            },
```
Unfortunately same issue.

Should I try to bypass the Broadcaster and call the `pusher->trigger` manually instead ?

**Comment by tjaydk** [comment]
My bad - just spotted it was first fixed in version `8.29.0` - I have now updated to latest `laravel/framework: v.8.32.1` and that did solve the issue. Thank you very much for the help. You can close the issue.

**Comment by elverkilde** [comment]
Thanks for confirming that, I'll update the docs for Laravel compatibility.
